// pg-client/lib/Database.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _Client = require('./Client');

var _Client2 = _interopRequireDefault(_Client);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Promise-based wrapper for `node-postgres`
 * https://github.com/brianc/node-postgres/wiki/pg
 */
/**
 * ES7 Async/Await Client for PostgreSQL
 * https://github.com/kriasoft/node-pg-client
 *
 * Copyright © 2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

function Database(pg) {
  this.pg = pg;
  this.defaults = pg.defaults;
  this.connect = this.connect.bind(this);
  this.end = this.end.bind(this);
}

Database.prototype.connect = function connect(connString, callback) {
  var _this = this;

  return new _bluebird2.default(function (resolve, reject) {
    var options = typeof connString === 'function' ? connString() : connString;
    _this.pg.connect(options, function (err, client, done) {
      if (err) {
        if (client) {
          // Remove the client from the connection pool.
          done(client);
        }
        reject(err);
      } else {
        callback(new _Client2.default(client)).then(function (result) {
          done();
          resolve(result);
        }, function (error) {
          done(client);
          reject(error);
        });
      }
    });
  });
};

Database.prototype.end = function end() {
  this.pg.end();
};

exports.default = Database;
